libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/RotateMapOverlay.kt

```diff
@@ -0,0 +1,174 @@
+package net.cyclestreets.views.overlay
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.content.pm.ActivityInfo
+import android.graphics.Canvas
+import android.graphics.drawable.Drawable
+import android.hardware.GeomagneticField
+import android.location.Location
+import android.location.LocationManager
+import android.support.design.widget.FloatingActionButton
+import android.util.Log
+import android.view.LayoutInflater
+import android.view.Surface
+import android.view.View
+import android.view.WindowManager
+import com.mikepenz.google_material_typeface_library.GoogleMaterial
+import com.mikepenz.iconics.IconicsDrawable
+import net.cyclestreets.util.Theme
+import net.cyclestreets.view.R
+import net.cyclestreets.views.CycleMapView
+import org.osmdroid.views.MapView
+import org.osmdroid.views.overlay.Overlay
+import org.osmdroid.views.overlay.compass.IOrientationConsumer
+import org.osmdroid.views.overlay.compass.IOrientationProvider
+import org.osmdroid.views.overlay.compass.InternalCompassOrientationProvider
+import org.osmdroid.views.overlay.mylocation.GpsMyLocationProvider
+import org.osmdroid.views.overlay.mylocation.IMyLocationConsumer
+import org.osmdroid.views.overlay.mylocation.IMyLocationProvider
+
+class RotateMapOverlay(private val mapView: CycleMapView)
+    : Overlay(), PauseResumeListener, IMyLocationConsumer, IOrientationConsumer
+{
+    private val rotateButton: FloatingActionButton
+    private val onIcon: Drawable
+    private val offIcon: Drawable
+    private val locationProvider: IMyLocationProvider
+    private val compassProvider: IOrientationProvider
+    private var rotate = false
+    private var gpsspeed = 0f
+    private var lat = 0f
+    private var lon = 0f
+    private var alt = 0f
+    private var timeOfFix: Long = 0
+    private var deviceOrientation = 0
+
+
+    init {
+        val context = mapView.context
+        onIcon = highlightIcon(context)
+        offIcon = lowlightIcon(context)
+
+        val rotateButtonView = LayoutInflater.from(context).inflate(R.layout.compassbutton, null)
+        rotateButton = rotateButtonView.findViewById(R.id.compass_button)
+        rotateButton.setOnClickListener { setRotation(!rotate) }
+
+        mapView.addView(rotateButtonView)
+        locationProvider = UseEverythingLocationProvider(context)
+        compassProvider = InternalCompassOrientationProvider(context)
+    }
+
+    private fun setRotation(state: Boolean) {
+        Log.d("LiveRide", "Setting map rotation to $state")
```

Comment:
Personally I prefer to put a static class `TAG` to avoid potentially repeating log strings like this.